tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
   return memoized_lookup;
 }
 
+// This function tries to transform
+//   dot(reshape(transpose(A)), Const) to
+//   dot(reshape(A), reshape(transpose(reshape(Const)))),
+// so that the reshape and transpose on the Const side can be constant folded.
+//
+// The basic idea is that since the accumulation in the dot operation is
+// associative, so as long as we permute the elements of the contracting
+// dimensions on both sides of the dot in the same way, the result of the
+// dot is not affected.
+StatusOr<HloInstruction*>
+AlgebraicSimplifierVisitor::OptimizeDotOfReorderContractingDims(
+    HloInstruction* dot) {
+  // This transformation assumes layout is not assigned yet.
+  if (options_.is_layout_sensitive()) {
+    return nullptr;
+  }
+
+  // Canonicalize dot(<constant>, rhs) to dot(rhs, <constant>) to make the
+  // remainder of this function easier.
+  auto dnums = dot->dot_dimension_numbers();
+  auto lhs_contracting_dims = dnums.lhs_contracting_dimensions();
+  auto rhs_contracting_dims = dnums.rhs_contracting_dimensions();
+  auto* lhs = dot->mutable_operand(0);
+  auto* rhs = dot->mutable_operand(1);
+  if (dot->operand(0)->IsConstant()) {
+    std::swap(lhs, rhs);
+    std::swap(lhs_contracting_dims, rhs_contracting_dims);
+  }
+
+  // Require single contracting dim to make the implementation easier to
+  // track contracting dims.
+  if (dnums.lhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+
+  // Pattern match Dot(reshape(transpose(input), constant))
+  HloInstruction* reshape;
+  HloInstruction* transpose;
+  HloInstruction* input;
+  HloInstruction* constant;
+  if (!Match(lhs,
+             m::Reshape(&reshape, m::Transpose(&transpose, m::Op(&input)))) ||
+      !Match(rhs, m::Constant(&constant))) {
+    return nullptr;
+  }
+
+  // Check if reshape squishes some dims into one dim, and that if this one
+  // dim is the dot's lhs contracting dim.
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
```

Comment:
Is this correct?

If the reshape squishes (say) 4 dimensions, then shouldn't `unmodified_dims.size() == lhs->shape().rank() - 4` (or possibly I'm off by one here, but I don't see how I'm off by 3...)?

Reply:
Note also that DimensionsUnmodifiedByReshape seems to allow us to create new 1-sized dimensions.  I am...not totally sure that does not break things?

Reply:
Here is the intent of this code: Since we are checking if the reshape squish multiple dimensions into 1, and now the lhs is the reshape op, then the number of unmodified dims should be the reshape result rank - 1, no matter how many input dims are squished. For example, if we have A = F32[a,b,c*d*e,f] reshape(F32[a,b,c,d,e,f] B), then the number of unmodified dims is 3 (a, b and f). If the number of unmodified dims is less than 3, then it means some dims other than the contracting dim is modified, then this reshape is not what we want.

Then we check if dim 2 (the lhs contracting dim) of A is unmodified, if it is unmodified, then it means that some dims other than contracting dim is modified, and that reshape is not what we want. If it is not in unmodifed, then it means this reshape squishes some dims into this contracting dim, and this reshape is what we want.

Reply:
Oh, I see, okay!  The thing I was missing is that `lhs->shape()` is *after* the reshape.

I'm wondering if we can make this more obvious somehow.  Perhaps it's just a matter of adding a brief comment on this `if` statement.  Or perhaps it's a matter of `s/lhs/reshape` (and similarly `s/lhs/transpose/` where appropriate.  WDYT?